### PR TITLE
[dynamo] Stop registering what a loaded nn.Module reads back as prunable guard state

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #197761
* #197760
* #197759
* #197758
* #197821
* #197835
* #197819
* #197833
* #197817
* __->__ #197752
* #197762
* #197756
* #197755
* #197754

**Problem.** The `nn.Module` branch of `GuardsStatePickler.reducer_override` registers every non-tensor, non-module, non-callable attribute the guard tree does not reach as prunable, including `_parameters`, `_buffers`, `_modules` and `_non_persistent_buffers_set`, which `nn.Module.__getattr__` indexes on every attribute miss, and `_all_weights`, which `RNNBase.__setstate__` reads at load. Today those survive only by accident: they are exact builtin containers, which the C pickler saves by type without ever consulting `reducer_override`.

**Fix.** The registration loop becomes `_prune_unguarded_attributes`, which skips the four bookkeeping containers, and it runs only for a module whose `__setstate__` is `nn.Module.__setstate__` (presence checks only) or for `DistributedDataParallel`, which is rebuilt through `nn.Module.__setstate__` explicitly; a module with its own `__setstate__` is pickled whole.

**Summary.** Pruning is sound when nothing reads the pruned attribute on the way back, and two kinds of module attribute are read back. `nn.Module.__getattr__` indexes `_parameters`, `_buffers` and `_modules` for every name outside `__dict__`, and `__setattr__`/`__delattr__` index all four containers on any assignment, so a loaded module whose bookkeeping had become the sentinel would raise `TypeError` on every parameter, buffer or submodule access instead of `AttributeError` on a genuine miss. A subclass with its own `__setstate__`, such as `RNNBase` indexing `self._all_weights[0][0]`, reads attributes no guard named, so any pruning at all is unsafe for it; the trade-off is that in-tree modules with a presence-check `__setstate__` (Conv, MultiheadAttention, the Transformer layers) are pickled whole, and an unpicklable bystander attribute on one of them bypasses the frame as it did before pruning existed. Both gates are latent at this commit because the C pickler's exact-container fast path never substituted those values anyway; the next commit closes that fast path with `persistent_id`, and without these gates it would turn every loaded module's attribute misses into `TypeError` and every saved LSTM into a load failure. The hook `OrderedDict`s are deliberately not on the list: pruning them is what keeps a module with a local-lambda hook serializable, at the accepted cost that `state_dict()` and `load_state_dict()` on a loaded module raise, since those iterate the hook dicts; ordinary attribute access and the guard checks do not.

```python
m = torch.nn.Linear(2, 2)
pickler = GuardsStatePickler({id(m): m}, {}, {}, {}, io.BytesIO())
pickler.dump({"m": m})
id(m.__dict__["_parameters"]) in pickler.missing_values
id(m.__dict__["_forward_hooks"]) in pickler.missing_values   # True, both

lstm = torch.nn.LSTM(4, 4)                 # RNNBase has its own __setstate__
pickler = GuardsStatePickler({id(lstm): lstm}, {}, {}, {}, io.BytesIO())
pickler.dump({"m": lstm})
id(lstm._all_weights) in pickler.missing_values   # parent: True; now: False (same for _flat_weights*)
```

`test_module_bookkeeping_containers_are_never_registered_as_pruned` and `test_module_with_its_own_setstate_is_not_pruned` assert on `missing_values` directly, with the container names spelled out rather than read from the constant under test, which is the only observable difference until the next commit. They fail on the parent.

```python
is_ddp = isinstance(obj, torch.nn.parallel.DistributedDataParallel)
if is_ddp or type(obj).__setstate__ is torch.nn.Module.__setstate__:
    self._prune_unguarded_attributes(obj)
```

`DistributedDataParallel.__setstate__` rebuilds the reducer and would fail without a default process group, which is why `_unpickle_ddp_module` already calls `torch.nn.Module.__setstate__` on it directly; that path reads nothing pruned, so DDP keeps being pruned.

Test Plan:

```
python test/dynamo/test_guard_serialization.py -k TestGuardsStatePickler
python test/dynamo/test_guard_serialization.py
```

The two new tests fail on the parent and pass here; the rest of the file is unchanged in behaviour.

Authored with an AI assistant.

Co-Authored-By: Claude Fable 5.1 <noreply@anthropic.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Thanks for the review. Disposition of each item against head `fe93ead21130`:

> the new test that guards the allowlist is tautological (it iterates the very frozenset it validates, so removing an entry can never fail it)

**Fixed** in fe93ead21130: the four names are spelled out in the test and the `_NN_MODULE_STATE_ATTRS` import is gone from the test file.

> two overlapping assertions, and the stronger one is brittle: ... Keep the targeted `assertNotIn` on `_all_weights` (and add `_flat_weights` / `_flat_weights_names`, which `__setstate__` also reads), drop the `== {}`.

**Fixed** in fe93ead21130 exactly as suggested.

> the comment's stated criterion does not match the allowlist's contents. ... The actual rule ... is: keep what the module reads on ordinary attribute access; prune the hook dicts deliberately, accepting that `state_dict()` on a loaded module raises

**Fixed** in fe93ead21130: the comment now states that rule (`__getattr__` indexes the three dicts, `__setattr__`/`__delattr__` index all four, hook dicts pruned deliberately with the `state_dict()`/`load_state_dict()` cost named); the PR body's Summary was corrected the same way.

> uses `is` for the `__setstate__` comparison while ... uses `==` for the structurally identical `__getstate__` comparison. Switch to `is`.

**Fixed** in fe93ead21130.

> the docstring ends with "(the caller checks that)" ... state the precondition as a precondition

**Fixed** in fe93ead21130.

> No round-trip test. ... A dump-then-load that touches `loaded.weight` and calls `loaded.state_dict()` would lock the invariant in this commit

**Declined**: at this commit a dump-then-load passes with or without the gates, because the C pickler's exact-container fast path never substitutes these values yet, so the round trip would pin nothing here. The round-trip tests (`test_loaded_nested_module_keeps_its_bookkeeping_containers`, `test_loaded_rnn_module_survives_its_own_setstate`) are in #197753, the commit that closes the fast path and where they fail without these gates.

> No test that DDP still prunes.

**Declined**: asserting on `missing_values` for DDP needs a live process group in a pickler-level test; the DDP arm is the two-token `is_ddp or` and the capture-level `test_ddp_module` exercises it end to end.

> No test for a presence-check-`__setstate__` module (Conv2d, Softmax).

**Declined** for now: whether those modules should be pruned is the open question below, and pinning either behaviour before it is decided would be premature.

> the `__setstate__` gate is considerably broader than the hazard ... an explicit deny-set (`RNNBase`, plus anything else audited as actually-reading) preserves the safety property with a fraction of the coverage loss

**Not changed, left for the author to decide.** The blanket gate is conservative by design: an out-of-tree `__setstate__` can read anything, and an in-tree allowlist has to be re-audited every time a new in-tree `__setstate__` appears. The coverage lost is narrower than it looks: on the parent, exact-container attributes of those modules were never pruned anyway (fast path), so the only new bypass is an unpicklable non-container bystander stored directly on a Conv/MHA/Transformer module. The PR body states the trade-off; switching to a deny-set is a design change I am not making unilaterally.

**Comment by bobrenjc93** [comment]
@claude
